packages-content-model/roosterjs-content-model-plugins/lib/autoFormat/keyboardListTrigger.ts

```diff
@@ -1,59 +1,55 @@
-    isBlockGroupOfType,
-    updateListMetadata,
-} from 'roosterjs-content-model-core';
-import type { IStandaloneEditor } from 'roosterjs-content-model-types';
+import { getSelectedSegmentsAndParagraphs } from 'roosterjs-content-model-core';
+import { setListStartNumber, setListStyle, setListType } from 'roosterjs-content-model-api';
+import type { ContentModelDocument, IStandaloneEditor } from 'roosterjs-content-model-types';
 
 /**
  * @internal
  */
-export default function keyboardListTrigger(
+export function keyboardListTrigger(
     editor: IStandaloneEditor,
     shouldSearchForBullet: boolean = true,
     shouldSearchForNumbering: boolean = true
 ) {
-    editor.formatContentModel((model, _) => {
+    editor.formatContentModel((model, _context) => {
         const listStyleType = getListTypeStyle(
             model,
             shouldSearchForBullet,
             shouldSearchForNumbering
         );
         if (listStyleType) {
+            const paragraph = getSelectedSegmentsAndParagraphs(model, false)[0][1];
+            if (paragraph) {
+                paragraph.segments.splice(0, 1);
+            }
             const { listType, styleType, index } = listStyleType;
-
-            const paragraphOrListItems = getOperationalBlocks(model, ['ListItem'], []);
-            paragraphOrListItems.forEach(({ block, parent }, itemIndex) => {
-                if (!isBlockGroupOfType(block, 'General')) {
-                    const blockIndex = parent.blocks.indexOf(block);
-                    const listLevel = createListLevel(listType, {
-                        startNumberOverride: index,
-                        direction: block.format.direction,
-                        textAlign: block.format.textAlign,
-                        marginTop: '0',
-                        marginBlockEnd: '0px',
-                        marginBlockStart: '0px',
-                    });
-                    updateListMetadata(listLevel, metadata => {
-                        return (metadata =
-                            listType == 'UL'
-                                ? {
-                                      unorderedStyleType: styleType,
-                                  }
-                                : {
-                                      orderedStyleType: styleType,
-                                  });
-                    });
-
-                    const listItem = createListItem([listLevel]);
-                    parent.blocks.splice(blockIndex, 1, listItem);
-                }
-            });
-
+            triggerList(editor, model, listType, styleType, index);
             return true;
         }
         return false;
     });
 }
+
+const triggerList = (
+    editor: IStandaloneEditor,
+    model: ContentModelDocument,
+    listType: 'OL' | 'UL',
+    styleType: number,
+    index?: number
+) => {
+    setListType(model, listType);
+    const isOrderedList = listType == 'OL';
+    if (index && isOrderedList) {
```

Comment:
what if index is 0?

Reply:
is that allowed?

Reply:
index 0 is not allowed

Reply:
Ok, in that case please add a comment to tell we don't expect index to be 0 here.